pod-operation/src/components/hall_effect.rs

```diff
@@ -0,0 +1,51 @@
+use ads1x1x::ic::{Ads1015, Resolution12Bit};
+use ads1x1x::interface::I2cInterface;
+use ads1x1x::mode::OneShot;
+use ads1x1x::ChannelSelection::{SingleA0, SingleA1, SingleA2, SingleA3};
+use ads1x1x::{Ads1x1x, DynamicOneShot, SlaveAddr, FullScaleRange};
+use nb::block;
+use rppal::i2c::I2c;
+
+fn voltage_to_current(voltage: i16) -> f32 {
+    // Convert voltage to f32 and scale it
+    let voltage = f32::from(voltage) / 1000.0;
+    
+    // Multiply voltage by 15.873 and subtract 39.683
+    let current = (voltage * 15.873 - 39.683) as f32;
+
+    // Print the voltage value
+    println!("Voltage: {}", voltage);
+    
+    // Return the current value
+    current
+}
+
+
+
+pub struct HallEffect {
+    ads1015: Ads1x1x<I2cInterface<I2c>, Ads1015, Resolution12Bit, OneShot>,
+}
+
+impl HallEffect {
+    pub fn new(device_address: SlaveAddr) -> Self {
+        let i2cdev = I2c::new().unwrap();
+        let mut adc = Ads1x1x::new_ads1015(i2cdev, device_address);
+        adc.set_full_scale_range(FullScaleRange::Within4_096V).unwrap();
+        HallEffect { ads1015: adc }
+    }
+
+    pub fn cleanup(self) {
+        self.ads1015.destroy_ads1015();
+    }
+
+    pub fn read_currents(&mut self) -> (f32, f32, f32, f32) {
+        let currents: [f32; 4] = [SingleA0, SingleA1, SingleA2, SingleA3]
+            .map(|channel| {
+                let voltage = block!(self.ads1015.read(channel)).unwrap() * 2;
```

Comment:
Question: is there a particular reason we're multiplying this voltage by 2 here?

Reply:
The voltages we're reading are greater than the maximum value provided by the 12-bit ADC of 2047 mV. Because of this, we changed the full scale range to a maximum of 4096, but since the ADC isn't able to output 4096 due to its 12-bit restriction, the actual value gets scaled down somewhat. Since the default range is 0 to 2048, the range right now doubles so the actual values outputted by the ADC are half what they should be. That's my understanding of it anyways. The documentation, at least from what I saw, doesn't really explain too much about this.

Reply:
Ahh, so for the FSR of ±4.096V, the LSB of the voltage value represents 2mV. Perhaps we'll want to extract this into a separate function so the intent of rescaling is clear?

Reply:
Non-critical, will revisit to perhaps include a comment explaining the FSR/LSB usage.